src/FSharp.Stats/Integration/Differentiation.fs

```diff
@@ -22,3 +26,66 @@ module Differentiation =
         let b  = (x1 - x0 - a*t1*t1)/t1
         (2.*a*t) + b
 
+    ///A two-point estimation is to compute the slope of a nearby secant line through two points.
+    ///This gives an approximations of f'(x) at x respectively to two points "x and x+h"/"x-h and x+h"(depending on the used algorithm) of the function f.
+    ///Choosing a small number h, h represents a small change in x, and it can be either positive or negative.
+    module TwoPointDifferentiation =
+    
+        //correcture of the Array.iterUntili function; here no boundarys are set to the lower and upper end
+        /// Iterates the data array beginning from the startIdx. 
+        /// The step size and direction are implied by magnitude and sign of stepSize. The function returns
+        /// the idx of the first value for which predicate returns true or the end/start of the collection
+        /// is reached (returning None). The predicate function takes the idx of the current value as an additional
+        /// parameter.
+        let private iterUntiliWithoutCutOff (predicate: int -> 'T -> bool) stepSize startIdx (arr: 'T []) =
+            let rec loop  (arr: 'T []) currentIdx =
+                if currentIdx < 0 then None
+                elif currentIdx > arr.Length-1 then None
+                else                                              
+                    match predicate currentIdx arr.[currentIdx] with 
+                    | true -> Some currentIdx   
+                    | _               -> loop arr (currentIdx+stepSize) 
+            loop arr startIdx 
+
+        ///Returns the approximation of f'(x) at x by calculating the two point differentiation.
+        ///h is the window for the difference calculation. f is the function for which to calculate numerical differentiation. x is the point at which the difference between "x and x+h"/"x-h and x+h" is calculated.
+        let differentiate h f x =
+            //error is proportional to h
+            let forwardDifference = 
+                ( (f (x+h)) - (f x) ) / h     
+
+            //error is proportional to h^2, so it decreases with this function for h<1
+            let centralDifference = 
+                ( (f (x+h)) - (f (x-h)) ) / (2.*h) 
+            if h >= 1. then forwardDifference else centralDifference
+
+        //if something is wrong with the following function, try to implement the function shown next: TODO: https://www.johndcook.com/NumericalODEStepSize.pdf
+        //source for the function below: http://math.bd.psu.edu/faculty/stevens/Old-Courses/MTHBD423-Fall2003/Notes-Handouts/ndiff.pdf
+        ///Returns the optimal size for h from all tested values in hArr. f is the function and x the point at which the numerical differentiation is calculated.
+        let optimalStepSize hArr f x = 
```

Comment:
```suggestion
        let private optimalStepSize hArr f x = 
```
I do not think this function should be public, i don not see any use of it outside of the context of the differentiation function